Capitolo Stime (#225)
Il-Libro-Open-Source/book: docs/it/stime.md

Nell'ambito informatico, la stima, può essere usata per dare una dimensione ad un task o ad un intero progetto, aiuta il business a delineare un'idea o a pianificare le attività ed è quindi piuttosto importante perché consente di prendere decisioni su come spendere le risorse; dobbiamo ricordarci infatti che lo sviluppo è parte del business, per molti versi lo stiamo facendo anche noi quando programmiamo.

Viene piuttosto naturale capire che, il cosiddetto Time To Market (tempo di arrivo su un certo mercato), può risultare determinante nella riuscita di un prodotto; impatta per forza di cose anche su come il business lo percepisce perché determina quando inizieranno a tornare i profitti da quello che di fatto è investimento. In poche parole il business prende decisioni su questo parametro che diventa perno, insieme ad altri fattori, per delineare le strategie aziendali. Pianificare in base ad una stima temporale non è l'unica maniera di gestire questo aspetto, ma è sicuramente il preponderante che quindi merita di essere prima di tutto compreso come processo.

> Cercando di comprender meglio la natura di quello che facciamo dobbiamo andare a definire cosa è il processo di produzione del software e quali incognite nasconde. Dal punto di vista della variabilità è delineabile come definito, statistico ed empirico; il tipo definito è il più semplice, non c'è alcuna sorpresa durante lo svolgimento, è qualcosa che abbiamo già fatto quindi non richiede alcuno sforzo perché sappiamo già prevederlo. L'esempio tipico è quello della configurazione di un servizio, l'installazione di un software e così via, non ha alcuna variabilità se non minima quindi qui la stima è deterministica e non comporta problematiche. Quello statistico è similare ma con variabilità più alta, qui si possono fare delle stime basate su esperienze precedenti, abbiamo una vasta gamma di pattern che si ripresentano e quindi predire un tempo di esecuzione è meno semplice del precedente ma sempre fattibile.

Reviewer: ```suggestion
Cercando di comprendere meglio la natura di quello che facciamo dobbiamo andare a definire cosa è il processo di produzione del software e quali incognite nasconde. Dal punto di vista della variabilità è delineabile come definito, statistico ed empirico; il tipo definito è il più semplice, non c'è alcuna sorpresa durante lo svolgimento, è qualcosa che abbiamo già fatto quindi non richiede alcuno sforzo perché sappiamo già prevederlo. L'esempio tipico è quello della configurazione di un servizio, l'installazione di un software e così via, non ha alcuna variabilità se non minima quindi qui la stima è deterministica e non comporta problematiche. Quello statistico è similare ma con variabilità più alta, qui si possono fare delle stime basate su esperienze precedenti, abbiamo una vasta gamma di pattern che si ripresentano e quindi predire un tempo di esecuzione è meno semplice del precedente ma sempre fattibile.
```

> Cercando di comprendere meglio la natura di quello che facciamo dobbiamo andare a definire cosa è il processo di produzione del software e quali incognite nasconde. Dal punto di vista della variabilità è delineabile come definito, statistico ed empirico; il tipo definito è il più semplice, non c'è alcuna sorpresa durante lo svolgimento, è qualcosa che abbiamo già fatto quindi non richiede alcuno sforzo perché sappiamo già prevederlo. L'esempio tipico è quello della configurazione di un servizio, l'installazione di un software e così via, non ha alcuna variabilità se non minima quindi qui la stima è deterministica e non comporta problematiche. Quello statistico è similare ma con variabilità più alta, qui si possono fare delle stime basate su esperienze precedenti, abbiamo una vasta gamma di pattern che si ripresentano e quindi predire un tempo di esecuzione è meno semplice del precedente ma sempre fattibile.

Reviewer 2: Questi primi due o tre punti mi sembrano troppo "ricchi" dal punto di vista lessicale, ho dovuto leggerlo un paio di volte per capire ahah magari riscriverei l'inizio in maniera più semplice. "Dal punto di vista della variabilità è delineabile come definito, statistico ed empirico", per esempio, potrebbe diventare "Possiamo definire principalmente tre modalità di produzione del software: Definito, statistico ed empirico" o qualcosa di simile, cosa ne pensi?

Author: @Cadienvan ho chiarito i due concetti chiarendoli con qualche esempio, scorre meglio? Grazie!

#### Isolare le criticità

Nella suddivisione si possono scorgere, come si è visto, criticità; sono le parti più complesse da stimare, quelle che ci pongono di fronte a situazioni nuove dove un pattern già conosciuto non è visibile o proprio è assente. Si possono adottare due strategie concatenate: la prima è quella di "assegnare ad una stima del tempo di stima": sembra un gioco di parole ma non lo è, se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo. Assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto di PoC: proof of concept. Di fatto è un micro task orientato alla creazione di uno o più proprietà del progetto finale che sono critiche e che, una volta sbrogliate, rendono tutto il processo di creazione più chiaro. Può anche essere utile mostrarlo, a volte può bastare provare se l'idea funziona a livello tecnico, ma a volte è possibile anche mostrarlo a chi poi prenderà decisioni perché dà immediatamente un'idea di dove si vuole arrivare, che prestazioni o di che interattività si parla.

#### Forbice

> L'espressione della stima è intuitivamente temporale in ore o giorni uomo, ma non è l'unica possibilità. In alcuni framework agili la sostituisce con delle unità adimensionali o addirittura con qualcosa di non numerico: l'esempio classico sono le taglie delle magliette. Questo approccio è tipico di quei team che lavorano per sessioni (chiamati a volte sprint) che durano da una settimana in su dove il gruppo sa che entreranno un certo numero di task con una determinata taglia, rimanere vaghi è un metodo per evitare due cose, la prima è che si faccia della matematica non opportuna sulle stime: la sequenza di come si svolge il lavoro è importante e sommarle ciecamente può comportare problemi nello svolgimento poi. La seconda è che si vuole semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.

Reviewer 2: ```suggestion
L'espressione della stima è intuitivamente temporale in ore o giorni uomo, ma non è l'unica possibilità. In alcuni framework agili questa viene sostituita con delle unità adimensionali o addirittura con qualcosa di non numerico: l'esempio classico sono le taglie delle magliette. Questo approccio è tipico di quei team che lavorano per sessioni (chiamati a volte sprint) che durano da una settimana in su dove il gruppo sa che entreranno un certo numero di task con una determinata taglia, rimanere vaghi è un metodo per evitare due cose, la prima è che si faccia della matematica non opportuna sulle stime: la sequenza di come si svolge il lavoro è importante e sommarle ciecamente può comportare problemi nello svolgimento poi. La seconda è che si vuole semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.
```

> L'espressione della stima è intuitivamente temporale in ore o giorni uomo, ma non è l'unica possibilità. In alcuni framework agili la sostituisce con delle unità adimensionali o addirittura con qualcosa di non numerico: l'esempio classico sono le taglie delle magliette. Questo approccio è tipico di quei team che lavorano per sessioni (chiamati a volte sprint) che durano da una settimana in su dove il gruppo sa che entreranno un certo numero di task con una determinata taglia, rimanere vaghi è un metodo per evitare due cose, la prima è che si faccia della matematica non opportuna sulle stime: la sequenza di come si svolge il lavoro è importante e sommarle ciecamente può comportare problemi nello svolgimento poi. La seconda è che si vuole semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.

Author: @serenasensini ho cercando di chiarire la questione "matematica sulle stime", fammi sapere se va meglio"

Se la stima invece è scritta sotto forma di tempo possiamo usare lo stratagemma di creare una forbice con un tempo minimo e massimo per lo svolgimento del lavoro, questo per conteggiare il rischio soprattutto di quelle parti che risultano nuove che potrebbero portare con sé delle criticità. Un altro modo è spesso quello di indicare una stima e poi aggiungere un margine, questo è forse il metodo più incerto perché il margine è spesso arbitrario; generalmente si usa Pareto aggiungendo il venti per cento a quanto ottenuto.

> #### Specifiche

Reviewer 2: Espanderei un pochino tutta questa sezione.

Parli di requisiti, dominio applicativo, andare alla fonte (Product Team / Business Owner di riferimento), cerimonie..

Sono cose che dal punto di vista di chi sa già cosa siano sono chiare, ma per il target di questo libro potrebbero essere cose nuove o comunque difficilmente comprensibili. Che ne pensi? Lo espanderei anche a 15-20 paragrafi se necessario, magari mettendo un cappello iniziale a tutto il capitolo dicendo che quando parliamo di stime dobbiamo considerare il contesto aziendale e che questo è composto da tanti fattori, tra cui....

Partendo sempre dal metodo empirico ci accorgiamo prima di tutto che l'incertezza comporta una cosa tanto banale quanto vera: sappiamo quanto ci abbiamo messo quando lo abbiamo fatto. Questo perché è solamente alla fine che è chiaro quello che si voleva produrre, in una parola serviva sapere quale era la definizione di fatto.

> Per saperlo è necessario avere delle specifiche ed è facile intuire che più sono precise più è semplice pianificare. Anche chi lavora da poco nel settore sa che sono complicate da ottenere e a volte fumose. Il primo consiglio qui è quello di studiare il dominio applicativo perché la già una nomenclatura delle cose da uno slancio non irrisorio. Poi è necessario parlare con chi ha questa conoscenza, andarle a cercare alla fonte e chiedere più chiarimenti possibili; anche in questo caso la scomposizione e l'iterazione sono utili. Ci sono delle cerimonie specifiche in alcuni framework e metodologie che possono venirci incontro es. event storming il cui scopo è proprio quello di chiarire i flussi e nomi del dominio il più possibile, qui la contaminazione col business è fondamentale per un prodotto di qualità e ridurre al minimo i fraintendimenti.

Reviewer 2: ```suggestion
Per saperlo è necessario avere delle specifiche ed è facile intuire che più sono precise più è semplice pianificare. Anche chi lavora da poco nel settore sa che sono complicate da ottenere e a volte fumose. Il primo consiglio qui è quello di studiare il dominio applicativo perché conoscere la nomenclatura e il perimetro delle cose da un vantaggio non indifferente. È necessario poi parlare con chi ha questa conoscenza, andarle a cercare alla fonte e chiedere più chiarimenti possibili; anche in questo caso la scomposizione e l'iterazione sono utili. Ci sono delle cerimonie specifiche in alcuni framework e metodologie che possono venirci incontro es. event storming il cui scopo è proprio quello di chiarire i flussi e nomi del dominio il più possibile, qui la contaminazione col business è fondamentale per un prodotto di qualità e ridurre al minimo i fraintendimenti.
```

#### Condivisione

La ricerca di pattern, abbiamo detto, è molto utile perché ci consente di riconoscere cose già viste o fatte; per quelle non conosciute il PoC e/o del tempo allocato per lo studio è un ottimo approccio. Resta però che così è vincolato alla nostra mera conoscenza dell'argomento, abbiamo un solo punto di vista che al più possiamo contaminare leggendo quello degli altri su articoli o libri ma l'opinione dal vivo racchiude molto più valore.

> Si può infrangere questa barriera effettuando dei meeting di condivisione e di stima delle entità dei problemi in gruppo; il problema visto da angolazioni differenti può svelare caratteristiche di cui non avevamo tenuto conto, anche solo un singolo confronto può portare grande valore ma più è di gruppo meglio è.

Reviewer 2: ```suggestion
Ma come possiamo ottenere contaminazione e condivisione in maniera efficace?

Si possono organizzare dei meeting di condivisione e di stima dei problemi; il problema visto da angolazioni differenti può svelare caratteristiche di cui non avevamo tenuto conto, e anche un singolo confronto può risultare fondamentale nel risolvere una problematica che sembrava insormontabile.
```

> In alcuni framework agili ci sono proprio dei momenti per questo tipo di approccio, per citiarne uno: "planning poker". Una gamification di questo processo che si svolge intorno ad un tavolo dove ogni uno, usando un mazzo di carte appositamente numerato (spesso usando Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltare il punto di vista degli altri e le loro considerazioni consentono di arrivare ad una convergenza appunto di una visione condivisa del problema e quindi ad una stima.

Reviewer: ```suggestion
In alcuni framework agili ci sono proprio dei momenti per questo tipo di approccio, per citarne uno: "planning poker". Una gamification di questo processo che si svolge intorno ad un tavolo dove ogni uno, usando un mazzo di carte appositamente numerato (spesso usando Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltare il punto di vista degli altri e le loro considerazioni consentono di arrivare ad una convergenza appunto di una visione condivisa del problema e quindi ad una stima.
```

> In alcuni framework agili ci sono proprio dei momenti per questo tipo di approccio, per citarne uno: "planning poker". Una gamification di questo processo che si svolge intorno ad un tavolo dove ogni uno, usando un mazzo di carte appositamente numerato (spesso usando Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltare il punto di vista degli altri e le loro considerazioni consentono di arrivare ad una convergenza appunto di una visione condivisa del problema e quindi ad una stima.

Reviewer 2: ```suggestion
In alcuni framework agili il problema viene risolto alla fonte tramite delle apposite cerimonie, per citarne uno: _Planning Poker_. Una _gamification_ di questo processo che si svolge intorno ad un tavolo dove ognuno, usando un mazzo di carte appositamente numerato (spesso usando la sequenza di Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltando il punto di vista degli altri e le loro considerazioni, consentendo di arrivare non solo ad una convergenza e ad una visione condivisa del problema, ma anche ad una stima.
```

Aggiungerei un "Come mai si usa la sequenza di Fibonacci?", che ne pensi?

Author: @Cadienvan rivisto il paragrafo rendendolo più descrittivo sia come procedimento che per la numerazione

> In alcuni framework agili ci sono proprio dei momenti per questo tipo di approccio, per citarne uno: "planning poker". Una gamification di questo processo che si svolge intorno ad un tavolo dove ogni uno, usando un mazzo di carte appositamente numerato (spesso usando Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltare il punto di vista degli altri e le loro considerazioni consentono di arrivare ad una convergenza appunto di una visione condivisa del problema e quindi ad una stima.

Reviewer 3: ```suggestion
In alcuni framework agili ci sono proprio dei momenti per questo tipo di approccio, per citarne uno: "planning poker". Una gamification di questo processo che si svolge intorno ad un tavolo dove ogni persona, usando un mazzo di carte appositamente numerato (spesso usando Fibonacci), dà la sua stima esprimendo il suo punto di vista. Si svolgono più mani proprio come a poker, ascoltare il punto di vista degli altri e le loro considerazioni consentono di arrivare ad una convergenza appunto di una visione condivisa del problema e quindi ad una stima.
```

#### MVP

> Il concetto di riduzione e scomposizione porta con sé un approccio naturale, cioè definire quali siano le parti fondamentali dell'applicativo. Si può introdurre un concetto semplice ma piuttosto efficiente ed efficace: dare una risposta alla domanda "quali sono le funzionalità minime per poter andare sul mercato?"; questo accorcia di molto il tempo di arrivo sul mercato perché elimina le funzioni non necessarie, semplifica la visione e lo sviluppo. Questo concetto è alla base del pensiero agile e va sotto il nome di Minimum Viable Product (MVP), abbassa il rischio rendendo più corto lo sviluppo, chiarisce meglio i goal del progetto e aiuta ancora una volta a scomporre le cose in elementi più piccoli o, se vogliamo, minimi perché il progetto abbia quel valore sul mercato che il business si aspetta.

Reviewer 2: Espanderei tutta la sezione dedicandogli qualche paragrafo in più, dall'introduzione del concetto di MVP al contesto e al risultato che questo comporta. Come sopra, è una lettura che risulterebbe sufficiente e sensata solo a chi già conosce il concetto - anche a grandi linee -, ma non al target tipico di questo contenuto. Terrei comunque il significato immutato, perché comunque è corretto, semplicemente lo espanderei per aiutare chi sente il termine per la prima, o seconda, volta.

Nell'ambito informatico, la stima può essere usata per dare una dimensione ad un task o ad un intero progetto, aiutando il business a delineare un'idea o a pianificare le attività ed è quindi piuttosto importante perché consente di prendere decisioni su come spendere le risorse. Dobbiamo ricordarci infatti che lo sviluppo è parte del business, e per molti versi lo stiamo facendo anche noi quando programmiamo.

Un parametro importante nello sviluppo di un progetto è il cosiddetto Time To Market (tempo di arrivo su un certo mercato), può risultare determinante nella riuscita di un prodotto; rappresenta quando riuscirete ad essere pronti per i clienti e quindi quando arriverete rispetto alla concorrenza. Il business lo tiene molto in considerazione perché è un buon indicatore di quando i profitti potrebbero iniziare ad arrivare cominciando, quindi, a rientrare degli investimenti. Questo parametro diventa perno, insieme ad altri fattori, per delineare le strategie aziendalie; delineare la pianificazione in base ad alla stima temporale non è l'unica maniera di gestire questo aspetto, ma è sicuramente spesso il preponderante.

> Ma come mai la stima e la pianificazione, nel nostro campo, è così complessa? La risposta sta nello comprendere, prima di tutto, quale è la natura del processo di creazione del software stesso, cerchiamo quindi di andare a definirnr la natura e quali incognite nasconda. I processi si categorizzano in base alla variabilità cioè di delle incognite che si incotrano durante il percorso; più è difficile predire quello che succederà durante il lavoro e più si dice che la variabilità cresce; è un poco come percorrere una strada e non sapere quanto traffico troveremo, se faremo quella strada molto spesso avremo comunque un'idea di cosa possa succedere e di quanto ci metteremo ad arrivare.

Reviewer 3: ```suggestion
Ma come mai la stima e la pianificazione, nel nostro campo, è così complessa? La risposta sta nel comprendere, prima di tutto, qual è lo scopo del processo di creazione del software stesso. Questo ci permette di definire meglio quelle variabili fino a quel momento incognite per cui le stime possono variare. 

I processi si categorizzano infatti base ad una serie di variabili, ossia delle incognite che si incontrano durante il percorso; più è difficile predire quello che succederà durante il lavoro e più si dice che la variabilità cresce; è un poco come percorrere una strada e non sapere quanto traffico troveremo, se faremo quella strada molto spesso avremo comunque un'idea di cosa possa succedere e di quanto ci metteremo ad arrivare.
```

Ho cercato di aggiustare un po' la frase, ma non mi torna il concetto dietro: cosa volevi dire? Le stime dipendono da delle variabili che all'inizio del processo sono incognite?

> Partendo da questo concetto il processo di produzione di un qualsiasi prodotto è delineabile come definito, statistico ed empirico; il tipo definito è il più semplice, non c'è alcuna sorpresa durante lo svolgimento, è qualcosa che abbiamo già fatto quindi non richiede alcuno sforzo perché sappiamo già prevederlo a monte. Nel nostro campo l'esempio tipico è quello della configurazione di un servizio, l'installazione di un software e così via, non ha alcuna variabilità se non minima quindi qui la stima è deterministica e non comporta problematiche.

Reviewer 3: ```suggestion
Partendo da questo concetto il processo di produzione di un qualsiasi prodotto può essere visto come definito, statistico ed empirico; il tipo "definito" è il più semplice, perché non c'è alcuna sorpresa durante lo svolgimento ed è qualcosa che abbiamo già fatto, quindi non richiede alcuno sforzo perché sappiamo già prevederlo a monte. Nel nostro campo l'esempio tipico è quello della configurazione di un servizio, l'installazione di un software e così via; si tratta di processi che non hanno alcuna variabilità se non minima quindi qui la stima è deterministica e non porta criticità particolari, se no quelle che riguardano l'adattamento allo specifico caso d'uso.
```

> Quello statistico è similare ma con variabilità più alta, significa che lo abbiamo già fatto ma sappiamo che potrebbero esserci dei problemi durante lo svolgimento; qui si possono fare delle stime basate, appunto, su esperienze precedenti, abbiamo una vasta gamma di pattern che si ripresentano e quindi predire un tempo di esecuzione è meno semplice del precedente ma sempre fattibile magari con una forbice.

Reviewer 3: ```suggestion
Quello "statistico" è simile al precedente, ma con variabilità più alta: significa che lo abbiamo già fatto ma sappiamo che potrebbero esserci dei problemi durante lo svolgimento. Qui si possono fare delle stime basate, per l'appunto, su esperienze precedenti, e per fortuna abbiamo una vasta gamma di pattern che possiamo sfruttare e che ci permettono di stabilire un tempo di esecuzione. In questo caso, si tratta di una stima meno semplice della precedente ma sempre fattibile, anche se magari con una "forbice" che definisce un intervallo entro cui muoversi.
```

> L'ultimo è quello più ostico e lo è perché percorriamo una strada non battuta: è quello che spesso ci si trova ad affrontare nel software che ha a che fare precisamente con ricerca e sviluppo; questo è di fatto il caso peggiore e richiede un approccio più complesso perché lo si conosce una volta che lo abbiamo fatto, non abbiamo un'esperienza specifica che lo possa descrivere. Dobbiamo cercare quindi di portarci verso gli altri due casi e diminuire l'incertezza mano mano per raffinazioni successiva, in maniera incrementale. In altre parole, dobbiamo delineare delle strategie.

Reviewer 3: ```suggestion
L'ultimo è quello più ostico e lo è perché percorriamo una strada non battuta: è quello più frequente e che ha a che fare precisamente con processi di ricerca e sviluppo. Questo è di fatto il caso peggiore e richiede un approccio più complesso, perché la reale stima la si conosce una volta che il processo giunge a termine, visto che non abbiamo un'esperienza specifica che lo possa descrivere. Dobbiamo cercare quindi di portarci verso gli altri due casi e diminuire l'incertezza mano mano per raffinazioni successiva, in maniera incrementale. In altre parole, dobbiamo delineare delle strategie.
```

Reviewer 3: Aggiungo una nota: rispetto a queste 3 definizioni, questa mi sembra quella che può portare più confusione. Ci può stare che una stima su un processo "non conosciuto" a priori sia difficile da fare, ma non credo riguardi i processi di ricerca e sviluppo (non sempre, almeno) e che la stima la si conosca solo alla fine del processo, mi sembra fuorviante, ma magari sbaglio... Hai della letteratura di riferimento?

Author: @serenasensini rileggendo noto che il mio approccio tende a ritornare molto su quello che è scrum e l'approccio/visione che ha delle cose. Qui la questione è che per sapere quanto ci vuole nella totalità è necessario definire bene cosa significa fatto ma, per farlo, sono necessarie più iterazioni per definire cosa significhi "fatto" per il committente. Qui per stima non intendo quella dell'iterazione singola ma di quella per arrivare alla conclusione. Mi rifaccio a quanto ho imparato tempo fa, ho trovato solo questo video che lo riassume ma, come dicevo, gira sul concetto di definizione di fatto per potare verso scrum e potrebbe essere effettivamente fuorviante.

[https://vimeo.com/176463126](https://vimeo.com/176463126)

Cosa ne pensi?
Sicuramente è da rivedere, se dici che non è troppo legato a scrum (mio dubbio) posso provare a riformulare.
Grazie

Reviewer 3: Sicuramente si vede l'impronta legata a Scrum, che può starci (visto il tema), ma va dettagliato: io credo che sia importante espandere questa sezione mostrando l'esempio  anche nel video, per portare più dettaglio. La cosa che mi preme è che sia semplice e contestualizzato anche per chi è a digiuno di Scrum! So che non è un compito facile, ma sono sicura tu possa riuscirci 🙌

Author: @serenasensini ho fatto una bella (spero) integrazione, posso chiederti di dare un occhio? Grazie!

Author: Ho aggiunto la descrizione dell'event storming, ho fatto una cosa leggera

#### Comprendere il processo

Quando si parla di stima si intende riuscire a delineare un confine temporale per la realizzazione di un'opera di qualche genere. Ci troviamo di fronte ad un tema controverso, e sicuramente chiedendo a dieci persone cosa rappresenta per loro una stima, probabilmente otterremo dieci risposte diverse con metodi o approcci anche molto distanti. Qui non vogliamo andare nel profondo di uno o più metodi, ma chiarire il tema per comprenderlo e dare degli strumenti e spunti di approfondimento.

> Nell'ambito informatico, la stima può essere usata per dare una dimensione ad un task o ad un intero progetto, aiutando il business a delineare un'idea o a pianificare le attività ed è quindi piuttosto importante perché consente di prendere decisioni su come spendere le risorse. Dobbiamo ricordarci infatti che lo sviluppo è parte del business, e per molti versi lo stiamo facendo anche noi quando programmiamo.

Reviewer 4: Non mi torna la frase "Dobbiamo ricordarci infatti che lo sviluppo è parte del business, e per molti versi lo stiamo facendo anche noi quando programmiamo.". Nel senso che leggo... "lo asviluppo è parte del business" e poi leggo "lo stiamo facendo anche noi quando programmiamo".

Cioè. Certo che chi programma sviluppa. Quell'anche mi disorienta.

Reviewer 2: Credo che il senso espresso fosse che "anche noi facciamo business quando programmiamo"

Author: @Cadienvan esattamente, @sensorario provo a riformularla

Author: Ho migliorato il fraseggio

Ma come mai la stima e la pianificazione, nel nostro campo, è così complessa? La risposta sta nel comprendere, prima di tutto, qual è lo scopo del processo di creazione del software stesso. Questo ci permette di definire meglio quelle variabili fino a quel momento incognite per cui le stime possono variare.

I processi si categorizzano infatti base ad una serie di variabili, ossia delle incognite che si incontrano durante il percorso; più è difficile predire quello che succederà durante il lavoro e più si dice che la variabilità cresce; è un poco come percorrere una strada e non sapere quanto traffico troveremo, se faremo quella strada molto spesso avremo comunque un'idea di cosa possa succedere e di quanto ci metteremo ad arrivare.

Per capire meglio quanto delicata sia questa tematica bisogna ricordarsi che il software è di fatto equiparabile alla produzione di un bene, è un prodotto che deve soddisfare le esigenze di uno o più clienti, un mercato.

> Il processo di creazione di un prodotto, dal punto di vista prettamente ingegneristico, è suddivisibile secondo questa categorizzazione: definito, statistico e empirico, va applicato quello migliore per la situazione in cui siamo.

Author: Hai ragione, rende meglio se metto un punto e proseguo con una nuova frase.

> Il processo di creazione di un prodotto, dal punto di vista prettamente ingengneristico, è suddivisibile in tre categorie: definito, statistico e empirico, va applicato quello migliore per la situazione in cui siamo.

Reviewer 4: ```suggestion
Il processo di creazione di un prodotto, dal punto di vista prettamente ingegneristico, è suddivisibile secondo questa categorizzazione: definito, statistico e empirico, va applicato quello migliore per la situazione in cui siamo.
```

Un esempio nel nostro campo questo potrebbe essere l'installazione di un sistema operativo, di un CMS; non ci aspettiamo grandi soprese da queste operazioni, al più possono chiederci qualche pacchetto accessorio ma nulla che possa impattare nè sui tempi nè sui risultati.

Il processo statistico, invece, ha una variabilità più alta cioè durante la sua esecuzione le cose possono andare diversamente da quanto previsto e questo impatta sui tempi e costi.
L'esempio classico è la produzione di un bene materiale dove per forza di cose, errori, difformità di materiali e altre variabili possono incidere sul risuotato che produciamo; ci troviamo di fronte ad un processo che può essere dominato con la statistica, cioè so che avrò un certo scarto percentuale.
Questa casistica è piuttosto rara in campo informatico ma comunque presente, per fare un esempio potremmo pensare all'installazione di un plugin su un CMS e alla sua configurazione che può variare da cliente a cliente.

> In questo caso non è detto che ci metteremo lo stesso quantitativo di tempo per ogni cliente, ecco qui che quindi è meglio esprimere la stima in una forbice perché abbiamo un'esperienza che ci dice che possiamo metterci da un minimo ad un massimo.

Author: Miglioro con un esempio esplicativo.

> Nei due precedenti processi ci accorgiamo presto che un elemento determinante è avere una specifica ben definita di cosa vogliamo ottenere, "cosa significa fatto?", "dove vogliamo arrivare?".

Reviewer 4: ```suggestion
Nei due precedenti processi, definito e statistico, ci accorgiamo presto che un elemento determinante è avere una specifica ben definita di cosa vogliamo ottenere, "cosa significa fatto?", "dove vogliamo arrivare?".
```

In questo contesto di incertezza nasce il processo empirico che ha preso piede negli ultimi venti anni, si tratta di abbracciare un po' questa incertenzza e procedere in maniera iterative.
Nei processi precedenti si nota che abbiamo un input, eseguiamo una lavorazione, emettiamo un output, a questo punto quello che abbiamo ottenuto è qualcosa che può andare bene o no; non abbiamo in mano specifiche molto precise (lo voglio lungo X e largo Y) e quindi il rischio di non soddisfare il cliente è alto.
Possiamo pensare di ragionare come nel caso di un pittore che deve creare un quadro che di per sè ha un soggetto e un tema ma non ha una definizione precisa del risultato finale.
Si procede prima con un disegno, che poi viene ripassato a tempera per poi venire ritoccato più volte per andare incontro alle esigenze del committente.
Il processo empirico è proprio questo, partire da una o più caratteristiche base, da uno scheletro, e agginugere elementi interagendo così da ottenere qualcosa che si adatta man mano che viene creato.

> Partire da elementi di base e poi costruire in manirea iterativa ci consente di scomporre tutto in elementi più piccoli e semplici, più predittibili e quindi più gestibili; questa scomposizione ci consente di riportarci ai due processi precedenti che possono essere predetti meglio e, obiettivo principale, automatizzati.

Author: Hai ragione, lo sottolineo come importante.

#### Scomposizione

> La prima tattica è quello di dividere il problema in problemi più semplic: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Author: Ottima osservazione, aggiungo un esempio

> La prima tattica è quello di dividere il problema in problemi più semplic: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Author: Vero, correggo.

> La prima tattica è quello di dividere il problema in problemi più semplic: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Reviewer 3: ```suggestion
La prima tattica è quello di dividere il problema in problemi più semplici: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.
```

> La prima tattica è quello di dividere il problema in problemi più semplic: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Reviewer 2: @allevo ti torna l'aggiustamento?

> La prima tattica è quello di dividere il problema in problemi più semplic: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono più difficoltà. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Reviewer 2: @allevo Anche qui ti torna?

> #### Isolare le criticità

Author: Qui mi sento meno convinto, in questo contesto tenderei a sovrapporli, cosa ne pensi @Cadienvan @serenasensini ?

> #### Isolare le criticità

Reviewer 3: Concordo con @eppak : "criticità" rappresenta bene il concetto che spiega poi dopo, perché la scomposizione in problemi più piccoli permette proprio di evitare inutili complessità

#### Forbice

L'espressione della stima è intuitivamente temporale in ore o giorni uomo, ma non è l'unica possibilità. In alcuni framework agili questa viene sostituita con delle unità adimensionali o addirittura con qualcosa di non numerico: l'esempio classico sono le taglie delle magliette. Questo approccio è tipico di quei team che lavorano per sessioni (chiamati a volte sprint) che durano da una settimana in su dove il gruppo sa che entreranno un certo numero di task con una determinata taglia, rimanere vaghi è un metodo per evitare due cose, la prima è che si faccia della matematica non opportuna sulle stime: la sequenza di come si svolge il lavoro è importante e sommarle ciecamente può comportare problemi nello svolgimento poi; capita spesso, infatti, che chi sovraintende il lavoro abbia bisogno di avere una scaletta (ancora meglio una data) e la cosa più immediata è sommare i tempi ma magari non è la solzione migliore perché ci sono dipendenze tra i lavori o si ha un'idea globale del lavoro ma non sufficientemente dettagliata e possono sorgere fraintendimenti.

> La seconda è che si vuole semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.

Author: Ottimo spunto, lo rivedo.

#### Comprendere il processo

Quando si parla di stima si intende riuscire a delineare un confine temporale per la realizzazione di un'opera di qualche genere. Ci troviamo di fronte ad un tema controverso, e sicuramente chiedendo a dieci persone cosa rappresenta per loro una stima, probabilmente otterremo dieci risposte diverse con metodi o approcci anche molto distanti. Qui non vogliamo andare nel profondo di uno o più metodi, ma chiarire il tema per comprenderlo e dare degli strumenti e spunti di approfondimento.

Nell'ambito informatico, la stima può essere usata per dare una dimensione ad un task o ad un intero progetto, aiutando il business a delineare un'idea o a pianificare le attività ed è quindi piuttosto importante perché consente di prendere decisioni su come spendere le risorse. Dobbiamo ricordarci infatti che lo sviluppo è parte del business, e lo stiamo facendo anche noi quando programmiamo.

> Un parametro importante nello sviluppo di un progetto è il cosiddetto Time To Market (tempo di arrivo su un certo mercato), può risultare determinante nella riuscita di un prodotto; rappresenta quando riuscirete ad essere pronti per i clienti e quindi quando arriverete rispetto alla concorrenza. Il business lo tiene molto in considerazione perché è un buon indicatore di quando i profitti potrebbero iniziare ad arrivare cominciando, quindi, a rientrare degli investimenti. Questo parametro diventa perno, insieme ad altri fattori, per delineare le strategie aziendalie; delineare la pianificazione in base ad alla stima temporale non è l'unica maniera di gestire questo aspetto, ma è sicuramente spesso il preponderante.

Reviewer 5: ```suggestion
Un parametro importante nello sviluppo di un progetto è il cosiddetto Time To Market (tempo di arrivo su un certo mercato), può risultare determinante nella riuscita di un prodotto; rappresenta quando riuscirete ad essere pronti per i clienti e quindi quando arriverete rispetto alla concorrenza. Il business lo tiene molto in considerazione perché è un buon indicatore di quando i profitti potrebbero iniziare ad arrivare cominciando, quindi, a rientrare degli investimenti. Questo parametro diventa perno, insieme ad altri fattori, per delineare le strategie aziendali; delineare la pianificazione in base ad alla stima temporale non è l'unica maniera di gestire questo aspetto, ma è sicuramente spesso il preponderante.
```

Il processo di creazione di un prodotto, dal punto di vista prettamente ingegneristico, è suddivisibile secondo questa categorizzazione: definito, statistico e empirico. Va applicato quello migliore per la situazione in cui siamo.

Il processo definito è quello più semplice, è qualcosa che abbiamo già fatto e di cui conosciamo bene le specifiche e il cui processo di esecuzione è ben definito; in questa sitazione conosco sia tempi che modi e non mi aspetto alcuna variabilità che in parole povere significa solo che non ci sono sorprese durante il processo, avviene sempre alla stessa maniera.
Un esempio nel nostro campo questo potrebbe essere l'installazione di un sistema operativo, di un CMS; non ci aspettiamo grandi soprese da queste operazioni, al più possono chiederci qualche pacchetto accessorio ma nulla che possa impattare nè sui tempi nè sui risultati.

Il processo statistico, invece, ha una variabilità più alta cioè durante la sua esecuzione le cose possono andare diversamente da quanto previsto e questo impatta sui tempi e costi.

> L'esempio classico è la produzione di un bene materiale dove per forza di cose, errori, difformità di materiali e altre variabili possono incidere sul risuotato che produciamo; ci troviamo di fronte ad un processo che può essere dominato con la statistica, cioè so che avrò un certo scarto percentuale.

Reviewer 5: ```suggestion
L'esempio classico è la produzione di un bene materiale dove per forza di cose, errori, difformità di materiali e altre variabili possono incidere sul risultato che produciamo; ci troviamo di fronte ad un processo che può essere dominato con la statistica, cioè so che avrò un certo scarto percentuale.
```

In questo caso non è detto che ci metteremo lo stesso quantitativo di tempo per ogni cliente, ecco qui che quindi è meglio esprimere la stima in una forbice perché abbiamo un'esperienza che ci dice che possiamo metterci da un minimo ad un massimo.

Nei due precedenti processi, definito e statistico, ci accorgiamo presto che un elemento determinante è avere una specifica ben definita di cosa vogliamo ottenere, "cosa significa fatto?", "dove vogliamo arrivare?".
Il cliente ha un'esigenza e spesso non è in grado di esprimerla da subito in termini chiari e precisi, in un contesto come meccanica ed elettronica è più facile perché il cliente è in grado di descrivre le specifiche con caratteristiche fisiche misurabili; un cliente che chiede un sito internet invece difficilmente potrà fare altrettanto.
Ad esempio possiamo trovarci di fronte a percezioni di un prodotto terzo e molto qui può fare l'analisi perché ci può chiarire meglio la specifica, la definizione di cosa si vuole ottenere.

> In questo contesto di incertezza nasce il processo empirico che ha preso piede negli ultimi venti anni, si tratta di abbracciare un po' questa incertenzza e procedere in maniera iterative.

Reviewer 5: ```suggestion
In questo contesto di incertezza nasce il processo empirico che ha preso piede negli ultimi venti anni, si tratta di abbracciare un po' questa incertezza e procedere in maniera iterativa.
```

Nei processi precedenti si nota che abbiamo un input, eseguiamo una lavorazione, emettiamo un output, a questo punto quello che abbiamo ottenuto è qualcosa che può andare bene o no; non abbiamo in mano specifiche molto precise (lo voglio lungo X e largo Y) e quindi il rischio di non soddisfare il cliente è alto.
Possiamo pensare di ragionare come nel caso di un pittore che deve creare un quadro che di per sè ha un soggetto e un tema ma non ha una definizione precisa del risultato finale.
Si procede prima con un disegno, che poi viene ripassato a tempera per poi venire ritoccato più volte per andare incontro alle esigenze del committente.
Il processo empirico è proprio questo, partire da una o più caratteristiche base, da uno scheletro, e agginugere elementi interagendo così da ottenere qualcosa che si adatta man mano che viene creato.

> Partire da elementi di base e poi costruire in manirea iterativa ci consente di scomporre tutto in elementi più piccoli e semplici, più predittibili e quindi più gestibili; questa scomposizione ci consente di riportarci ai due processi precedenti che possono essere predetti meglio e, obiettivo importante, automatizzati magari per un riuso in altri ambiti.

Reviewer 5: ```suggestion
Partire da elementi di base e poi costruire in maniera iterativa ci consente di scomporre tutto in elementi più piccoli e semplici, più predittibili e quindi più gestibili; questa scomposizione ci consente di riportarci ai due processi precedenti che possono essere predetti meglio e, obiettivo importante, automatizzati magari per un riuso in altri ambiti.
```

Per fare un esempio, partendo dal classico to do list, potremmo mettere in risalto autenticazione, le operazioni di base di visione/creazione/modifica/cancellazione di un elemento, a questo aggiungiamo che il nostro committente voglia anche condivisione dell'item e magari memo sul calendario. Ora possiamo individuare l'autenticazione e le operazioni di base come piuttosto ben definite, ma condivisione e integrazione col calendario? Cominciamo a porci delle domande per esempio "condividere con chi" o "quale calendario usare" (Google?). Vediamo quindi emergere dei macro blocchi e degli elementi da chiarire per i quali individuare anche degli esperti sia tecnologici che non che possono aiutarci a farci un'idea di cosa vogliamo ottenere; in tutto questo, quando troviamo uno scoglio nella scomposizione, possiamo in prima battuta impegnare uno slot di tempo per lo studio ben definito il classico "time boxed" oltre il quale non andare e dopo il quale va cercato aiuto.

#### Isolare le criticità

> Nella suddivisione si possono scorgere, come si è visto, criticità; sono le parti più complesse da stimare, quelle che ci pongono di fronte a situazioni nuove dove un pattern già conosciuto non è visibile o proprio è assente. Si possono adottare due strategie concatenate: la prima è quella di "assegnare ad una stima del tempo di stima": sembra un gioco di parole ma non lo è. Se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo. Assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto di PoC: proof of concept. Di fatto è un micro task orientato alla creazione di uno o più proprietà del progetto finale che sono critiche e che, una volta sbrogliate, rendono tutto il processo di creazione più chiaro. Può anche essere utile mostrarlo, a volte può bastare provare se l'idea funziona a livello tecnico, ma a volte è possibile anche mostrarlo a chi poi prenderà decisioni perché dà immediatamente un'idea di dove si vuole arrivare, che prestazioni o di che interattività si parla.

Reviewer 5: suggerisco di fare un elenco puntato con le due strategie in modo da dividere bene il testo (anche visualmente). Inoltre si eviterebbe il ":" dentro un altro ":".

> #### Forbice

Reviewer 5: ```suggestion
#### Unità di misura
```
O simili.

- Semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.
  Se la stima invece è scritta sotto forma di tempo possiamo usare lo stratagemma di creare una forbice con un tempo minimo e massimo per lo svolgimento del lavoro, questo per conteggiare il rischio soprattutto di quelle parti che risultano nuove che potrebbero portare con sé delle criticità. Un altro modo è spesso quello di indicare una stima e poi aggiungere un margine, questo è forse il metodo più incerto perché il margine è spesso arbitrario; generalmente si usa Pareto aggiungendo il venti per cento a quanto ottenuto.

> #### Specifiche

Reviewer 5: ```suggestion
#### L'impatto delle specifiche sulla stima
```
che te ne pare ?

# Le stime temporali di progetto

## Introduzione

Per gestire correttamente un progetto o un singolo task è fondamentale stimare in modo quantitativo le risorse necessarie a realizzare un prodotto che offra funzionalità di valore per chi lo commissiona (gli stakeholder). Quando parliamo di risorse, il pensiero va subito (spesso in modo implicito) al tempo, cioè quanto occorre per arrivare sul mercato o per ottenere una versione funzionante. Tuttavia, il tempo è solo uno degli elementi da considerare in una stima progettuale: di seguito vediamo alcuni esempi senza entrare troppo nel dettaglio.

- **Costo di funzionamento**: riguarda ciò che serve perché il prodotto possa “girare”. Può essere suddiviso in hardware e software minimi o, se il progetto è cloud-native, nelle risorse allocate dal provider. Vanno inoltre stimati i costi di monitoraggio, manutenzione, aggiornamenti di sicurezza e supporto. Questi fattori, fondamentali in fase di business plan, incidono sul costo finale del servizio.

> - **Capacity**: indica quanto lavoro è realizzabile in un determinato periodo. Serve a capire quante persone (e con quali competenze) sono necessarie per completare il progetto e in che tempi è possibile raggiungere il mercato. Questo aspetto impatta sulla fattibilità e sulla possibilità di occupare una posizione strategica.

Reviewer 5: Citerei anche la disponibilità delle persone: magari ora non ci sono persone disponibili perché già occupate su altri progetti e ci si può occupare del nuovo solo tra tot tempo.
Ovvero: la capacity non é solo numerica sulle persone ma anche temporale poiché va capita la disponibilità delle stesse.

Author: Coerente, aggiunto.

> - **Competenze**: sviluppare un e-commerce è ben diverso dal creare una piattaforma di streaming. Occorre valutare se il team disponga di tutte le conoscenze necessarie (ad esempio, l’uso di un database non relazionale per dati particolari). Se tali competenze mancano, si dovrà scegliere se formare internamente il personale o rivolgersi a consulenti esterni; entrambe le opzioni influiscono sui costi di realizzazione.

Reviewer 5: Poiché si é fatto l'esempio di e-commerce e piattaforma di streaming, continuerei con gli stessi contesti anche nell'esempio di conoscenze necessarie. Esempi: SEO, analitici vs scale di streaming, player di video, metadati come sottotitoli e simili.
Alcune competenze chiaramente possono essere condivise, come pratiche di sviluppo, search, campagne pubblicitarie

Author: Modificato usando esempi correlati.

> A complemento delle stime quantitative, esistono poi le stime qualitative, che considerano aspetti non funzionali come le prestazioni attese dal sistema, i requisiti di sicurezza, la compatibilità con altri sistemi o la scalabilità richiesta. Questi elementi completano il quadro di ciò che serve per portare a termine il progetto.

Reviewer 5: Toglierei la questione delle prestazioni e della scalabilità che sono misurabili per le quali si possono fare stime quantitative.

Author: Non mi trovi completamente allineato però concordo che la scalabilità possa dare un'idea sbagliata, quindi è stata rimossa. Sulle performance ho chiarito che le intendo come percepite dall'utente, esperienza d'uso.

Ma come mai la stima e la pianificazione, nel nostro campo, è così complessa? La risposta sta nel comprendere, prima di tutto, qual è lo scopo del processo di creazione del software stesso. Questo ci permette di definire meglio quelle variabili fino a quel momento incognite per cui le stime possono variare.

I processi si categorizzano infatti base ad una serie di variabili, ossia delle incognite che si incontrano durante il percorso; più è difficile predire quello che succederà durante il lavoro e più si dice che la variabilità cresce; è un poco come percorrere una strada e non sapere quanto traffico troveremo, se faremo quella strada molto spesso avremo comunque un'idea di cosa possa succedere e di quanto ci metteremo ad arrivare.

Per capire meglio quanto delicata sia questa tematica bisogna ricordarsi che il software è di fatto equiparabile alla produzione di un bene, è un prodotto che deve soddisfare le esigenze di uno o più clienti, un mercato.
Il processo di creazione di un prodotto, dal punto di vista prettamente ingegneristico, è suddivisibile secondo questa categorizzazione: definito, statistico e empirico. Va applicato quello migliore per la situazione in cui siamo.

> Il processo definito è quello più semplice, è qualcosa che abbiamo già fatto e di cui conosciamo bene le specifiche e il cui processo di esecuzione è ben definito; in questa sitazione conosco sia tempi che modi e non mi aspetto alcuna variabilità che in parole povere significa solo che non ci sono sorprese durante il processo, avviene sempre alla stessa maniera.

Reviewer 6: Typo

```suggestion
Il processo definito è quello più semplice, è qualcosa che abbiamo già fatto e di cui conosciamo bene le specifiche e il cui processo di esecuzione è ben definito; in questa situazione conosco sia tempi che modi e non mi aspetto alcuna variabilità che in parole povere significa solo che non ci sono sorprese durante il processo, avviene sempre alla stessa maniera.
```

> Un esempio nel nostro campo questo potrebbe essere l'installazione di un sistema operativo, di un CMS; non ci aspettiamo grandi soprese da queste operazioni, al più possono chiederci qualche pacchetto accessorio ma nulla che possa impattare nè sui tempi nè sui risultati.

Reviewer 6: Mi suona meglio così

```suggestion
Un esempio, nel nostro campo, potrebbe essere l'installazione di un sistema operativo, di un CMS; non ci aspettiamo grandi soprese da queste operazioni, al più possono chiederci qualche pacchetto accessorio ma nulla che possa impattare né sui tempi né sui risultati.
```

L'esempio classico è la produzione di un bene materiale dove per forza di cose, errori, difformità di materiali e altre variabili possono incidere sul risultato che produciamo; ci troviamo di fronte ad un processo che può essere dominato con la statistica, cioè so che avrò un certo scarto percentuale.
Questa casistica è piuttosto rara in campo informatico ma comunque presente, per fare un esempio potremmo pensare all'installazione di un plugin su un CMS e alla sua configurazione che può variare da cliente a cliente; possiamo immaginare che un plugin di personalizzazione grafica vari ad esempio la livrea colori e magari un logo.
In questo caso non è detto che ci metteremo lo stesso quantitativo di tempo per ogni cliente, ecco qui che quindi è meglio esprimere la stima in una forbice perché abbiamo un'esperienza che ci dice che possiamo metterci da un minimo ad un massimo.

Nei due precedenti processi, definito e statistico, ci accorgiamo presto che un elemento determinante è avere una specifica ben definita di cosa vogliamo ottenere, "cosa significa fatto?", "dove vogliamo arrivare?".

> Il cliente ha un'esigenza e spesso non è in grado di esprimerla da subito in termini chiari e precisi, in un contesto come meccanica ed elettronica è più facile perché il cliente è in grado di descrivre le specifiche con caratteristiche fisiche misurabili; un cliente che chiede un sito internet invece difficilmente potrà fare altrettanto.

Reviewer 6: Typo

```suggestion
Il cliente ha un'esigenza e spesso non è in grado di esprimerla da subito in termini chiari e precisi, in un contesto come meccanica ed elettronica è più facile perché il cliente è in grado di descrivere le specifiche con caratteristiche fisiche misurabili; un cliente che chiede un sito internet invece difficilmente potrà fare altrettanto.
```

Ad esempio possiamo trovarci di fronte a percezioni di un prodotto terzo e molto qui può fare l'analisi perché ci può chiarire meglio la specifica, la definizione di cosa si vuole ottenere.

In questo contesto di incertezza nasce il processo empirico che ha preso piede negli ultimi venti anni, si tratta di abbracciare un po' questa incertezza e procedere in maniera iterativa.
Nei processi precedenti si nota che abbiamo un input, eseguiamo una lavorazione, emettiamo un output, a questo punto quello che abbiamo ottenuto è qualcosa che può andare bene o no; non abbiamo in mano specifiche molto precise (lo voglio lungo X e largo Y) e quindi il rischio di non soddisfare il cliente è alto.

> Possiamo pensare di ragionare come nel caso di un pittore che deve creare un quadro che di per sè ha un soggetto e un tema ma non ha una definizione precisa del risultato finale.

Reviewer 6: ```suggestion
Possiamo pensare di ragionare come nel caso di un pittore che deve creare un quadro che di per sé ha un soggetto e un tema ma non ha una definizione precisa del risultato finale.
```

Si procede prima con un disegno, che poi viene ripassato a tempera per poi venire ritoccato più volte per andare incontro alle esigenze del committente.
Il processo empirico è proprio questo, partire da una o più caratteristiche base, da uno scheletro, e agginugere elementi interagendo così da ottenere qualcosa che si adatta man mano che viene creato.

> Partire da elementi di base e poi costruire in maniera iterativa ci consente di scomporre tutto in elementi più piccoli e semplici, più predittibili e quindi più gestibili; questa scomposizione ci consente di riportarci ai due processi precedenti che possono essere predetti meglio e, obiettivo importante, automatizzati magari per un riuso in altri ambiti.

Reviewer 6: Mi suona come un inglesismo forzato

```suggestion
Partire da elementi di base e poi costruire in maniera iterativa ci consente di scomporre tutto in elementi più piccoli e semplici, più prevedibili e quindi più gestibili; questa scomposizione ci consente di riportarci ai due processi precedenti che possono essere predetti meglio e, obiettivo importante, automatizzati magari per un riuso in altri ambiti.
```

## Scomposizione

> La prima tattica è quello di dividere il problema in problemi più semplici: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui ci sono possono nascondersi delle difficoltà, ma soprattuttto sono quelle che hanno una maggiore incertezza nella stima. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.

Reviewer 6: ```suggestion
La prima tattica è quella di dividere il problema in problemi più semplici: un task o un progetto diventano via via più raffinati e più piccoli e al tempo stesso più semplici da stimare ma anche più facili da confinare come problematica. Come in precedenza, si cerca di trovare un pattern conosciuto e partire con quello. Questo vuol dire gestire un lavoro più piccolo, il che ci consente di comprenderlo meglio, visto che il cervello umano non è in grado di tenere sotto controllo una moltitudine di elementi, ma ci consente di avere una visione prima di insieme e poi del dettaglio, così da rendere le cose più facili da dominare. La riduzione rende anche chiare altre due cose, ovvero le criticità: parliamo di quelle parti meno esplorate e in cui possono nascondersi delle difficoltà, ma soprattuttto sono quelle che hanno una maggiore incertezza nella stima. Mette in luce anche le dipendenze, cioè quali siano le fondamentali e quali siano le parti accessorie, cominciando anche a delineare una sequenza.
```

> Per fare un esempio, partendo dal classico to do list, potremmo mettere in risalto autenticazione, le operazioni di base di visione/creazione/modifica/cancellazione di un elemento, a questo aggiungiamo che il nostro committente voglia anche condivisione dell'item e magari memo sul calendario. Ora possiamo individuare l'autenticazione e le operazioni di base come piuttosto ben definite, ma condivisione e integrazione col calendario? Cominciamo a porci delle domande per esempio "condividere con chi" o "quale calendario usare" (Google?). Vediamo quindi emergere dei macro blocchi e degli elementi da chiarire per i quali individuare anche degli esperti sia tecnologici che non che possono aiutarci a farci un'idea di cosa vogliamo ottenere; in tutto questo, quando troviamo uno scoglio nella scomposizione, possiamo in prima battuta impegnare uno slot di tempo per lo studio ben definito il classico "time boxed" oltre il quale non andare e dopo il quale va cercato aiuto.

Reviewer 6: ```suggestion
Per fare un esempio, partendo dal classico to do list, potremmo mettere in risalto l'autenticazione, le operazioni di base di visione/creazione/modifica/cancellazione di un elemento, a questo aggiungiamo che il nostro committente voglia anche la condivisione dell'elemento e magari il memo sul calendario. Ora possiamo individuare l'autenticazione e le operazioni di base come ben definite, ma condivisione e integrazione col calendario? Cominciamo a porci delle domande per esempio "condividere con chi" o "quale calendario usare" (Google?). Vediamo quindi emergere dei macro blocchi e degli elementi da chiarire per i quali individuare anche degli esperti sia tecnologici che non che possono aiutarci a farci un'idea di cosa vogliamo ottenere; in tutto questo, quando troviamo uno scoglio nella scomposizione, possiamo in prima battuta impegnare uno slot di tempo per lo studio ben definito, il classico "time boxed" oltre il quale non andare e dopo il quale va cercato aiuto.
```

## Isolare le criticità

Nella suddivisione si possono scorgere, come si è visto, criticità; sono le parti più complesse da stimare, quelle che ci pongono di fronte a situazioni nuove dove un pattern già conosciuto non è visibile o proprio è assente. Si possono adottare due strategie concatenate:

> - Assegnare ad una stima del tempo di stima: sembra un gioco di parole ma non lo è. Se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo; assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto successivo.

Reviewer 6: ```suggestion
- Assegnare ad una stima, del tempo di stima: sembra un gioco di parole ma non lo è. Se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo; assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto successivo.
```

> - La prima è quella di "assegnare ad una stima del tempo di stima", sembra un gioco di parole ma non lo è. Se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo; assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto successivo.

Reviewer 2: ```suggestion
- Assegnare ad una stima del tempo di stima: sembra un gioco di parole ma non lo è. Se un tema è complesso e non scomponibile e ha bisogno di essere studiato è necessario prendersi il tempo per farlo; assegnare a questo task una stima per consentirci di avere le idee più chiare ci da la maniera di introdurre il concetto successivo.
```

> - PoC: proof of concept, di fatto è un micro task orientato alla creazione di uno o più proprietà del progetto finale che sono critiche e che, una volta sbrogliate, rendono tutto il processo di creazione più chiaro. Può anche essere utile mostrarlo, a volte può bastare provare se l'idea funziona a livello tecnico, ma a volte è possibile anche mostrarlo a chi poi prenderà decisioni perché dà immediatamente un'idea di dove si vuole arrivare, che prestazioni o di che interattività si parla.

Reviewer 2: ```suggestion
- PoC: Proof of Concept, di fatto è un micro task orientato alla creazione di uno o più proprietà del progetto finale che sono critiche e che, una volta sbrogliate, rendono tutto il processo di creazione più chiaro. Può anche essere utile mostrarlo, a volte può bastare provare se l'idea funziona a livello tecnico, ma a volte è possibile anche mostrarlo a chi poi prenderà decisioni perché dà immediatamente un'idea di dove si vuole arrivare, che prestazioni o di che interattività si parla.
```

### Unità di misura

L'espressione della stima è intuitivamente temporale in ore o giorni uomo, ma non è l'unica possibilità. In alcuni framework agili questa viene sostituita con delle unità adimensionali o addirittura con qualcosa di non numerico: l'esempio classico sono le taglie delle magliette. Questo approccio è tipico di quei team che lavorano per sessioni (chiamati a volte sprint) che durano da una settimana in su dove il gruppo sa che entreranno un certo numero di task con una determinata taglia, rimanere vaghi è un metodo per evitare raggiungere due obiettivi:

- Evitare che si faccia della matematica non opportuna sulle stime: la sequenza di come si svolge il lavoro è importante e sommarle ciecamente può comportare problemi nello svolgimento poi; capita spesso, infatti, che chi sovraintende il lavoro abbia bisogno di avere una scaletta (ancora meglio una data) e la cosa più immediata è sommare i tempi di stima. Purtroppo però questa scelta potrebbe non essere la soluzione migliore perché ci sono dipendenze tra i lavori o ci sono aree ancora troppo poco chiare per avere una stima affidabile.

> - Semplificare il lavoro di stima evitando di dare un numero in ore e quindi si sa che ce la si farà in una sessione ma si evita di approfondire quanto anche per non creare fraintendimenti.

Reviewer 6: ```suggestion
- Semplificare il lavoro di stima, evitando di dare un numero in ore: così si sa che ce la si farà in una sessione, ma si evita di approfondire quanto, anche per non creare fraintendimenti.
```

#### MVP

> Il concetto di riduzione e scomposizione porta con sé un approccio naturale, cioè definire quali siano le parti fondamentali dell'applicativo. Si può introdurre un concetto semplice ma piuttosto efficiente ed efficace: dare una risposta alla domanda "quali sono le funzionalità minime per poter andare sul mercato?"; questo accorcia di molto il tempo di arrivo sul mercato perché elimina le funzioni non necessarie, semplifica la visione e lo sviluppo. Questo concetto è alla base del pensiero agile e va sotto il nome di Minimum Viable Product (MVP), abbassa il rischio rendendo più corto lo sviluppo, chiarisce meglio i goal del progetto e aiuta ancora una volta a scomporre le cose in elementi più piccoli o, se vogliamo, minimi perché il progetto abbia quel valore sul mercato che il business si aspetta.

Reviewer 5: Parere personale: l'mvp permette di focalissare l'attenzione su poche cose, generando il valore massimo possibile. Le funzionalità accessorie rimangono fuori scope.
Metterei il focus su questo

Author: Il tratto che mi indichi sull'MVP a me fila, lo lascerei così come prima versione, potrei aggiungere la questione del valore massimo possibile col minimo possibile. Come ti suona?
